torch/optim/optimizer.py

```diff
@@ -404,9 +405,15 @@ def zero_grad(self, set_to_none: bool = False):
                             else:
                                 p.grad.requires_grad_(False)
                             if (not foreach or p.grad.is_sparse):
-                                p.grad.zero_()
+                                if p.grad.device.type == 'cuda':
```

Comment:
Note that they need to be on the same device idx as well. Using the per_device_and_dtype_grads will fix this,

Reply:
Hi @albanD, could you please approve the running workflows for the new commit?

Reply:
I don't think you're handling sparse properly by doing this.
As mentioned above, removing the constraint that the `foreach` flag is set and ensuring that the device is cuda sounds like a better approach.